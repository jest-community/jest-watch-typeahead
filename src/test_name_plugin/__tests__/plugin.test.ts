let pluginTester: typeof import('../../test_utils/pluginTester').default = null;
let TestNamePlugin: typeof import('../plugin').default = null;
let KEYS: typeof import('jest-watcher').KEYS = null;

const testResults = [
  {
    testResults: [
      { title: 'foo 1', fullName: 'some description foo 1' },
      { title: 'bar 1', fullName: 'some description bar 1' },
    ],
  },
  {
    testResults: [
      { title: 'foo 2', fullName: 'other description foo 2' },
      { title: 'bar 2', fullName: 'other description bar 2' },
    ],
  },
];

beforeAll(async () => {
  TestNamePlugin = (await import('../plugin')).default;
  pluginTester = (await import('../../test_utils/pluginTester')).default;
  KEYS = (await import('jest-watcher')).KEYS;
});

it('shows the correct initial state', async () => {
  const { stdout, updateConfigAndRun, plugin, type } =
    pluginTester(TestNamePlugin);

  const runPromise = plugin.run({}, updateConfigAndRun);
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();
  type(KEYS.ENTER);

  await runPromise;
});

it('shows the correct message when there are no cached tests', async () => {
  const { stdout, updateConfigAndRun, plugin, type } =
    pluginTester(TestNamePlugin);

  const runPromise = plugin.run({}, updateConfigAndRun);
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();
  type('t', KEYS.ENTER);

  await runPromise;
});

it('can use arrows to select a specific test', async () => {
  const { stdout, hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(TestNamePlugin);

  hookEmitter.onTestRunComplete({ testResults });
  const runPromise = plugin.run({}, updateConfigAndRun);
  stdout.write.mockReset();
  type('f', KEYS.ARROW_DOWN, KEYS.ARROW_DOWN, KEYS.ENTER);
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();

  await runPromise;

  expect(updateConfigAndRun).toHaveBeenCalledWith({
    mode: 'watch',
    testNamePattern: '^other description foo 2$',
  });
});

it('can select a pattern that matches multiple tests', async () => {
  const { stdout, hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(TestNamePlugin);

  hookEmitter.onTestRunComplete({ testResults });
  const runPromise = plugin.run({}, updateConfigAndRun);
  stdout.write.mockReset();
  type('f', 'o', KEYS.ENTER);
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();

  await runPromise;

  expect(updateConfigAndRun).toHaveBeenCalledWith({
    mode: 'watch',
    testNamePattern: 'fo',
  });
});

it('can select a pattern that matches a describe block', async () => {
  const { stdout, hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(TestNamePlugin);

  hookEmitter.onTestRunComplete({ testResults });
  const runPromise = plugin.run({}, updateConfigAndRun);
  stdout.write.mockReset();
  type('s', 'o', KEYS.ENTER);
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();

  await runPromise;

  expect(updateConfigAndRun).toHaveBeenCalledWith({
    mode: 'watch',
    testNamePattern: 'so',
  });
});

it('can select a pattern that includes a regexp special character', async () => {
  const { hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(TestNamePlugin);

  hookEmitter.onTestRunComplete({
    testResults: [
      {
        testResults: [
          { title: 'bracket', fullName: 'bracket description (foo)' },
        ],
      },
    ],
  });
  const runPromise = plugin.run({}, updateConfigAndRun);

  type('b', 'r', KEYS.ARROW_DOWN, KEYS.ENTER);

  await runPromise;

  expect(updateConfigAndRun).toHaveBeenCalledWith({
    mode: 'watch',
    testNamePattern: '^bracket description \\(foo\\)$',
  });
});

it('can configure the key and prompt', async () => {
  const { plugin } = pluginTester(TestNamePlugin, {
    config: {
      key: 'l',
      prompt: 'have a custom prompt',
    },
  });

  expect(plugin.getUsageInfo()).toEqual({
    key: 'l',
    prompt: 'have a custom prompt',
  });
});

it('test matching is case insensitive', async () => {
  const { stdout, hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(TestNamePlugin);

  hookEmitter.onTestRunComplete({ testResults });
  const runPromise = plugin.run({}, updateConfigAndRun);
  type('f');
  stdout.write.mockReset();
  type('O');
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();
  type(KEYS.ENTER);
  await runPromise;
});

it("selected pattern doesn't include trimming dots", async () => {
  const { hookEmitter, updateConfigAndRun, plugin, type } = pluginTester(
    TestNamePlugin,
    {
      stdout: { columns: 30 },
    },
  );

  hookEmitter.onTestRunComplete({
    testResults: [
      {
        testResults: [
          {
            title: 'trimmed long',
            fullName: 'long test name, gonna need trimming',
          },
        ],
      },
    ],
  });
  const runPromise = plugin.run({}, updateConfigAndRun);

  type('t', 'r', 'i', 'm', 'm', KEYS.ARROW_DOWN, KEYS.ENTER);
  await runPromise;

  expect(updateConfigAndRun).toHaveBeenCalledWith({
    mode: 'watch',
    testNamePattern: '^me, gonna need trimming$',
  });
});
