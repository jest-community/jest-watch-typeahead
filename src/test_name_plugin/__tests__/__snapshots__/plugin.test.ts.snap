// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can select a pattern that matches a describe block 1`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
<dim> pattern ›</> s
<saveCursorPosition>
<moveCursorToColumn1>


 Pattern matches 4 tests from <yellow>cached</> test suites

 <dim>›</> </>s</><dim>ome description foo 1</>

 <dim>›</> </>s</><dim>ome description bar 1</>

 <dim>›</> <dim>other de</></>s</><dim>cription foo 2</>

 <dim>›</> <dim>other de</></>s</><dim>cription bar 2</>
<moveCursorToRow6Column13>
<restoreCursorPosition>
<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
<dim> pattern ›</> so
<saveCursorPosition>
<moveCursorToColumn1>


 Pattern matches 2 tests from <yellow>cached</> test suites

 <dim>›</> </>so</><dim>me description foo 1</>

 <dim>›</> </>so</><dim>me description bar 1</>
<moveCursorToRow6Column14>
<restoreCursorPosition>
"
`;

exports[`can select a pattern that matches multiple tests 1`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
<dim> pattern ›</> f
<saveCursorPosition>
<moveCursorToColumn1>


 Pattern matches 2 tests from <yellow>cached</> test suites

 <dim>›</> <dim>some description </></>f</><dim>oo 1</>

 <dim>›</> <dim>other description </></>f</><dim>oo 2</>
<moveCursorToRow6Column13>
<restoreCursorPosition>
<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
<dim> pattern ›</> fo
<saveCursorPosition>
<moveCursorToColumn1>


 Pattern matches 2 tests from <yellow>cached</> test suites

 <dim>›</> <dim>some description </></>fo</><dim>o 1</>

 <dim>›</> <dim>other description </></>fo</><dim>o 2</>
<moveCursorToRow6Column14>
<restoreCursorPosition>
"
`;

exports[`can use arrows to select a specific test 1`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
<dim> pattern ›</> f
<saveCursorPosition>
<moveCursorToColumn1>


 Pattern matches 2 tests from <yellow>cached</> test suites

 <dim>›</> <dim>some description </></>f</><dim>oo 1</>

 <dim>›</> <dim>other description </></>f</><dim>oo 2</>
<moveCursorToRow6Column13>
<restoreCursorPosition>
<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
<dim> pattern ›</> f
<saveCursorPosition>
<moveCursorToColumn1>


 Pattern matches 2 tests from <yellow>cached</> test suites

 <dim>›</> <black><backgroundYellow>some description foo 1</></>

 <dim>›</> <dim>other description </></>f</><dim>oo 2</>
<moveCursorToRow6Column13>
<restoreCursorPosition>
<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
<dim> pattern ›</> f
<saveCursorPosition>
<moveCursorToColumn1>


 Pattern matches 2 tests from <yellow>cached</> test suites

 <dim>›</> <dim>some description </></>f</><dim>oo 1</>

 <dim>›</> <black><backgroundYellow>other description foo 2</></>
<moveCursorToRow6Column13>
<restoreCursorPosition>
"
`;

exports[`shows the correct initial state 1`] = `
"<hideCursor>
<clearTerminal>

<bold>Pattern Mode Usage</>
 <dim>› Press</> Esc <dim>to exit pattern mode.</>
 <dim>› Press</> Enter <dim>to filter by a tests regex pattern.</>


<showCursor>
<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
<dim> pattern ›</> 
<saveCursorPosition>
<moveCursorToColumn1>


 <italic><yellow>Start typing to filter by a test name regex pattern.</></>
<moveCursorToRow6Column12>
<restoreCursorPosition>
"
`;

exports[`shows the correct message when there are no cached tests 1`] = `
"<hideCursor>
<clearTerminal>

<bold>Pattern Mode Usage</>
 <dim>› Press</> Esc <dim>to exit pattern mode.</>
 <dim>› Press</> Enter <dim>to filter by a tests regex pattern.</>


<showCursor>
<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
<dim> pattern ›</> 
<saveCursorPosition>
<moveCursorToColumn1>


 <italic><yellow>Start typing to filter by a test name regex pattern.</></>
<moveCursorToRow6Column12>
<restoreCursorPosition>
"
`;

exports[`test matching is case insensitive 1`] = `
"<eraseLine>
<moveCursorToColumn1>
<eraseScreenDown>
<dim> pattern ›</> fO
<saveCursorPosition>
<moveCursorToColumn1>


 Pattern matches 2 tests from <yellow>cached</> test suites

 <dim>›</> <dim>some description </></>fo</><dim>o 1</>

 <dim>›</> <dim>other description </></>fo</><dim>o 2</>
<moveCursorToRow6Column14>
<restoreCursorPosition>
"
`;
